index/scorch/segment/zap/segment.go

```diff
@@ -295,10 +295,17 @@ var visitDocumentCtxPool = sync.Pool{
 // VisitDocument invokes the DocFieldValueVistor for each stored field
 // for the specified doc number
 func (s *SegmentBase) VisitDocument(num uint64, visitor segment.DocumentFieldValueVisitor) error {
-	// first make sure this is a valid number in this segment
 	if num < s.numDocs {
 		vdc := visitDocumentCtxPool.Get().(*visitDocumentCtx)
+		defer visitDocumentCtxPool.Put(vdc)
+		return s.visitDocument(vdc, num, visitor)
+	}
+	return nil
+}
 
+func (s *SegmentBase) visitDocument(vdc *visitDocumentCtx, num uint64, visitor segment.DocumentFieldValueVisitor) error {
+	// first make sure this is a valid number in this segment
+	if num < s.numDocs {
```

Comment:
Do we need to check the docNum again?  Seems like the contract could be that it's always check one place or the other.

Reply:
Yes, as this internal API can be invoked directly or via the public one above.

Reply:
chatted briefly with steve, since external method always calls internal one, we could ditch the extra check in the outer one.  yes, it might mean an extra call to get/return from the sync.pool, but an invalid doc num is an exceptional case anyway, who care if it is slightly slower.  keeping the regular case clearer and with one less branch seems more important.

Reply:
Done.